beacon-chain/cache/registration.go

```diff
@@ -0,0 +1,67 @@
+package cache
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+)
+
+// RegistrationCache is used to store the cached results of an Validator Registration request.
+// beacon api /eth/v1/validator/register_validator
+type RegistrationCache struct {
+	sync.RWMutex
+	indexToRegistration map[primitives.ValidatorIndex]*ethpb.ValidatorRegistrationV1
+}
+
+// NewRegistrationCache initializes the map and underlying cache.
+func NewRegistrationCache() *RegistrationCache {
+	return &RegistrationCache{
+		indexToRegistration: make(map[primitives.ValidatorIndex]*ethpb.ValidatorRegistrationV1),
+	}
+}
+
+// GetRegistrationByIndex returns the registration by index in the cache and also removes items in the cache if expired.
+func (regCache *RegistrationCache) GetRegistrationByIndex(id primitives.ValidatorIndex) (*ethpb.ValidatorRegistrationV1, error) {
+	regCache.RLock()
+	defer regCache.RUnlock()
+	v, ok := regCache.indexToRegistration[id]
+	if !ok {
+		return nil, errors.Wrapf(ErrNotFoundRegistration, "validator id %d", id)
+	}
+	if timeStampExpired(v.Timestamp) {
+		delete(regCache.indexToRegistration, id)
+		return nil, errors.Wrapf(ErrNotFoundRegistration, "validator id %d", id)
+	}
```

Comment:
This is a mutating operation. You will need to use a write lock if you want to modify the map in this method